### PR TITLE
PLT-322 Updated Direct Messages menu

### PR DESCRIPTION
**Pull request by hmhealey**
Does a bunch of things:
1) Adds Preferences table which is used to store each users' preferences, separate from the user object. Each preference has a category, name, an optional alternate id, and a value. The alt id can be used for things like assigning a preference to a specific channel or user.
2) Changes the Direct Messages list on the left sidebar to be controlled by a preference instead of populated based on who you've received messages from.
3) Adds the ability to add/remove people from your Direct Messages list

### CODE REVIEW

**Comment by hmhealey** [comment]
I need to make some changes because we have no generic way to handle upserting and the way that I was doing it causes Postgres to panic.

**Comment by hmhealey** [comment]
Postgres has been fixed although I had to write some queries by hand since Postgres actually has no method for doing an insert with an "ON DUPLICATE" update like MySQL does

File: api/preference.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/gorilla/mux"
+	"github.com/mattermost/platform/model"
+	"net/http"
+)
+
+func InitPreference(r *mux.Router) {
+	l4g.Debug("Initializing preference api routes")
+
+	sr := r.PathPrefix("/preferences").Subrouter()
+	sr.Handle("/set", ApiAppHandler(setPreferences)).Methods("POST")
```
Comment (coreyhulen) [inline api/preference.go]: I think everywhere else use use create vs set

File: api/preference.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	sr.Handle("/{category:[A-Za-z0-9_]+}/{name:[A-Za-z0-9_]+}", ApiAppHandler(getPreferencesByName)).Methods("GET")
+}
+
+func setPreferences(c *Context, w http.ResponseWriter, r *http.Request) {
+	var preferences []*model.Preference
+
+	decoder := json.NewDecoder(r.Body)
```
Comment (coreyhulen) [inline api/preference.go]: Write a Preferences object which is type Preference [] Preference with json func's.  See Audits.go

File: api/preference.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	if result := <-Srv.Store.Preference().SaveOrUpdate(preferences...); result.Err != nil {
+		c.Err = result.Err
+		return
+	}
+
+	w.Write([]byte("true"))
+}
+
+func getPreferencesByName(c *Context, w http.ResponseWriter, r *http.Request) {
```
Comment (coreyhulen) [inline api/preference.go]: It may be more obvious as I read more, but why isn't this ByCategory vs ByName?  I want to return all in 1 swoop vs multiple calls.

File: api/preference.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		if id == userId {
+			continue
+		}
+
+		profile := profiles[id]
+
+		preference := &model.Preference{
+			UserId:   userId,
+			Category: model.PREFERENCE_CATEGORY_DIRECT_CHANNELS,
+			Name:     model.PREFERENCE_NAME_SHOW,
```
Comment (coreyhulen) [inline api/preference.go]: hmm..I thought userId, cat, name was considered unique?  Name feels like another category.

File: model/preference.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+func (o *Preference) ToJson() string {
+	b, err := json.Marshal(o)
+	if err != nil {
+		return ""
+	} else {
+		return string(b)
+	}
+}
+
+func PreferenceListToJson(o []*Preference) string {
```
Comment (coreyhulen) [inline model/preference.go]: Move to Preferences.go

File: store/sql_preference_store.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+	go func() {
+		storeChannel <- s.save(s.GetMaster(), preference)
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func (s SqlPreferenceStore) save(queryable Queryable, preference *model.Preference) StoreResult {
```
Comment (coreyhulen) [inline store/sql_preference_store.go]: why all the crazy duplicate methods?

File: store/sql_preference_store.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+		result.Err = model.NewAppError("SqlPreferenceStore.Update", "We couldn't update the preference",
+			"user_id="+preference.UserId+", category="+preference.Category+", name="+preference.Name+", alt_id="+preference.AltId+", "+err.Error())
+	} else {
+		result.Data = count
+	}
+
+	return result
+}
+
+func (s SqlPreferenceStore) SaveOrUpdate(preferences ...*model.Preference) StoreChannel {
```
Comment (coreyhulen) [inline store/sql_preference_store.go]: Is there any case where I need to diff between save/update as a user of SqlPreferenceStore ?  if the answer is no, then I would remove Save() and Update() and just rename SaveOrUpdate() to Save().  It doesn't look like Save() or Update is used?

File: store/sql_store.go [hunk 8]
```diff
@@ -611,3 +619,11 @@ func decrypt(key []byte, cryptoText string) (string, error) {
 
 	return fmt.Sprintf("%s", ciphertext), nil
 }
+
+// Interface for both gorp.DbMap and gorp.Transaction to allow code for one to be reused with the other
+type Queryable interface {
```
Comment (coreyhulen) [inline store/sql_store.go]: Is this necessary?  Isn't Queryable all internal to Preferences?  can't you just use interface vs the unnecessary abstraction.

File: web/react/utils/async_client.jsx [hunk 9]
```diff
@@ -637,3 +637,56 @@ export function getMyTeam() {
         }
     );
 }
+
+export function getDirectChannels() {
```
Comment (coreyhulen) [inline web/react/utils/async_client.jsx]: It's this really getDirectChannelPreferences() ?

**Comment by coreyhulen** [comment]
Also please be careful with your overuse of varargs.  Looks like it most places []*model.Preference would suffice.

File: api/preference.go (hunk 1 above)
Comment (hmhealey) [inline api/preference.go]: Does create make sense there when setPreferences either creates or updates?

File: api/preference.go (hunk 3 above)
Comment (hmhealey) [inline api/preference.go]: Yep, I can add a separate ByCategory api call in addition to the existing ByName call. I only needed ByName for this so I only actually implemented that.

File: store/sql_preference_store.go (hunk 7 above)
Comment (hmhealey) [inline store/sql_preference_store.go]: I can remove those. Now that I think of it, I think SqlChannelStore.Save might also do an upsert as well, but it can do it the trial and error way since it isn't wrapped in a transaction at all

File: store/sql_store.go (hunk 8 above)
Comment (hmhealey) [inline store/sql_store.go]: It's internal now, but I figured it might be useful in the future if we ended up doing more things in transactions. And I can try just doing it with just a generic interface{}, but I'm not sure if we lose compile time type checking or anything that way.

File: web/react/utils/async_client.jsx (hunk 9 above)
Comment (hmhealey) [inline web/react/utils/async_client.jsx]: Good point, I'll change that

File: api/preference.go (hunk 2 above)
Comment (hmhealey) [inline api/preference.go]: Can do

File: api/preference.go (hunk 4 above)
Comment (hmhealey) [inline api/preference.go]: UserId, Category, Name, AltId is unique. I agree that category may not entirely serve a purpose at this point because of that so we might want to reconsider removing it.

File: api/preference.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/gorilla/mux"
+	"github.com/mattermost/platform/model"
+	"net/http"
+)
+
+func InitPreference(r *mux.Router) {
+	l4g.Debug("Initializing preference api routes")
+
+	sr := r.PathPrefix("/preferences").Subrouter()
+	sr.Handle("/save", ApiAppHandler(savePreferences)).Methods("POST")
```
Comment (hmhealey) [inline api/preference.go]: These need UserRequiredHandler

File: api/preference.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+func getPreferenceCategory(c *Context, w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	category := params["category"]
+
+	if result := <-Srv.Store.Preference().GetCategory(c.Session.UserId, category); result.Err != nil {
+		c.Err = result.Err
+	} else {
+		data := result.Data.(model.Preferences)
+
+		if len(data) == 0 {
```
Comment (hmhealey) [inline api/preference.go]: Refactor this into some transform function

File: model/preference.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	if len(o.Value) > 128 {
+		return NewAppError("Preference.IsValid", "Value is too long", "value="+o.Value)
+	}
+
+	return nil
+}
+
+func IsPreferenceCategoryValid(category string) bool {
```
Comment (hmhealey) [inline model/preference.go]: Remove constant validation

File: model/preference.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+package model
+
+import (
+	"encoding/json"
+	"io"
+)
+
+const (
+	PREFERENCE_CATEGORY_DIRECT_CHANNEL_SHOW = "direct_channel_show"
+	PREFERENCE_CATEGORY_TEST                = "test" // do not use, just for testing uniqueness while there's only one real category
```
Comment (hmhealey) [inline model/preference.go]: Remove these

File: store/sql_preference_store.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+func (s SqlPreferenceStore) Get(userId string, category string, name string) StoreChannel {
+	storeChannel := make(StoreChannel)
+
+	go func() {
+		result := StoreResult{}
+
+		var preferences model.Preferences
+
+		if _, err := s.GetReplica().Select(&preferences,
```
Comment (hmhealey) [inline store/sql_preference_store.go]: Switch to SelectOne

File: store/sql_preference_store_test.go [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+// See License.txt for license information.
+
+package store
+
+import (
+	"github.com/mattermost/platform/model"
+	"testing"
+)
+
+func TestPreferenceSave(t *testing.T) {
```
Comment (hmhealey) [inline store/sql_preference_store_test.go]: Add tests for getting empty category and preference

File: web/react/stores/preference_store.jsx [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+        this.emitChange = this.emitChange.bind(this);
+        this.addChangeListener = this.addChangeListener.bind(this);
+        this.removeChangeListener = this.removeChangeListener.bind(this);
+
+        this.handleEventPayload = this.handleEventPayload.bind(this);
+        this.dispatchToken = AppDispatcher.register(this.handleEventPayload);
+    }
+
+    getKey(category, name) {
```
Comment (hmhealey) [inline web/react/stores/preference_store.jsx]: Make these private

File: web/react/components/sidebar.jsx [hunk 17]
```diff
@@ -322,7 +304,35 @@ export default class Sidebar extends React.Component {
...
         });
     }
-    createChannelElement(channel, index) {
+
+    handleLeaveDirectChannel(channel) {
+        if (!this.isLeaving.get(channel.id)) {
+            this.isLeaving.set(channel.id, true);
+
+            const preference = PreferenceStore.setPreference(Constants.Preferences.CATEGORY_DIRECT_CHANNEL_SHOW, channel.teammate_id, 'false');
+            AsyncClient.savePreferences(
```
Comment (hmhealey) [inline web/react/components/sidebar.jsx]: Just change this to Client.savePreferences and remove the success and error parameters from AsyncClient

**Comment by hmhealey** [comment]
Merging since we did a group review
